src/engine/sched.c

```diff
@@ -1544,6 +1548,89 @@ sched_start_cycle(struct sched_data *data, ABT_pool *pools)
 	}
 }
 
+/*
+ * FIXME: Replace following hack with Argobots APIs (if proper APIs are
+ * exported in later Argobots version, like an API of getting function
+ * address from ABT_unit.)
+ */
+
+/* Hack start */
+struct ABTI_thread_dup;
+
+struct ABTI_thread_dup {
+	struct ABTI_thread_dup *p_prev;
+	struct ABTI_thread_dup *p_next;
+	int is_in_pool;
+	uint32_t type;
+	ABT_unit unit;
+	void *p_last_xstream;
+	void *p_parent;
+	void (*f_thread)(void *);
+	void *p_arg;
+};
+
+struct sched_unit {
+	uint64_t	su_start;
+	void		*su_func_addr;
+};
+
+static inline bool
+watchdog_enabled(struct dss_xstream *dx)
+{
+	/* Enable watchdog for sys xstream only */
+	return sched_unit_runtime_max != 0 && dx->dx_xs_id == 0;
+}
+
+static void
+sched_watchdog_prep(struct dss_xstream *dx, ABT_unit unit,
+		    struct sched_unit *su)
+{
+	struct ABTI_thread_dup	*td = (struct ABTI_thread_dup *)unit;
+
+	if (!watchdog_enabled(dx))
+		return;
+
+	su->su_start = daos_getntime_coarse() / NSEC_PER_MSEC;
```

Comment:
I recall you added daos_getmtime_coarse recently.

Reply:
yeah, but that patch hasn't been landed yet.